server/server.js

```diff
@@ -1,15 +1,940 @@
+require("dotenv").config();
 const express = require("express");
+const cors = require("cors");
+const { Pool } = require("pg");
 const app = express();
 const port = process.env.PORT || 5000;
 
-app.listen(port, () => console.log(`Listening on port ${port}`));
+app.use(cors());
+app.use(express.json());
 
-// Store and retrieve your videos from here
-// If you want, you can copy "exampleresponse.json" into here to have some data to work with
-let videos = [];
+// PostgreSQL database connection
+const db = new Pool({
+  user: process.env.USERNAME,
+  host: process.env.HOSTNAME,
+  database: process.env.DATABASE_NAME,
+  password: process.env.PASSWORD,
+  port: process.env.DB_PORT, // test to add
+  // port: 5432,
+});
+
+db.connect()
+  .then(() => {
+    console.log("Connected to PostgreSQL database");
+  })
+  .catch((err) => {
+    console.error("Error connecting to PostgreSQL database:", err);
+  });
 
-// GET "/"
 app.get("/", (req, res) => {
-  // Delete this line after you've confirmed your server is running
   res.send({ express: "Your Backend Service is Running" });
 });
+
+app.get("/videos", async (req, res) => {
+  try {
+    const order = req.query.order || "desc";
+
+    let query = "SELECT * FROM videos";
+    if (order === "asc") {
+      query += " ORDER BY rating ASC";
+    } else {
+      query += " ORDER BY rating DESC";
+    }
+
+    const result = await db.query(query);
+    res.json(result.rows);
+  } catch (error) {
+    console.error("Error fetching videos:", error);
+    res.status(500).json({ error: "Internal Server Error" });
+  }
+});
+
+app.post("/videos", async (req, res) => {
+  try {
+    const newVideo = req.body;
+    const timestamp = new Date().toISOString();
+
+    const query =
+      "INSERT INTO videos (title, url, rating, timestamp) VALUES ($1, $2, $3, $4) RETURNING id";
+    const values = [newVideo.title, newVideo.url, 0, timestamp];
+
+    const result = await db.query(query, values);
+    const insertedVideoId = result.rows[0].id;
+
+    res.json({ id: insertedVideoId });
+  } catch (error) {
+    console.error("Error adding video:", error);
+    res.status(500).json({ error: "Internal Server Error" });
+  }
+});
+
+app.post("/videos/:id/upvote", async (req, res) => {
+  try {
+    const id = parseInt(req.params.id);
+
+    const query = "UPDATE videos SET rating = rating + 1 WHERE id = $1";
+    const values = [id];
+
+    await db.query(query, values);
+    res.json({ message: "Upvoted successfully" });
+  } catch (error) {
+    console.error("Error upvoting video:", error);
+    res.status(500).json({ error: "Internal Server Error" });
+  }
+});
+
+app.post("/videos/:id/downvote", async (req, res) => {
+  try {
+    const id = parseInt(req.params.id);
+
+    const query =
+      "UPDATE videos SET rating = GREATEST(rating - 1, 0) WHERE id = $1";
+    const values = [id];
+
+    await db.query(query, values);
+    res.json({ message: "Downvoted successfully" });
```

Comment:
With this kind of endpoint where the client hasn't supplied the value to set, we often would return the new value in the response, so that the UI can display the new value (which would factor in that other people may have changed the rating too), and means we can avoid replicating the logic on the client side to do the 0-clamping.